publishing/s3publisher.py

```diff
@@ -0,0 +1,180 @@
+
+        contents = response.get('Contents')
+        if not contents:
+            return remote_objects
+
+        for response_obj in contents:
+            # remove the site_prefix from the key
+            filename = remove_prefix(response_obj['Key'], site_prefix)
+
+            # remove initial slash if present
+            filename = remove_prefix(filename, '/')
+
+            # the etag comes surrounded by double quotes, so remove them
+            md5 = response_obj['ETag'].replace('"', '')
+
+            site_obj = SiteObject(filename=filename, md5=md5,
+                                  site_prefix=site_prefix)
+            remote_objects.append(site_obj)
+
+        results_truncated = response['IsTruncated']
+        if results_truncated:
+            continuation_token = response['NextContinuationToken']
+
+    return remote_objects
+
+
+def publish_to_s3(directory, base_url, site_prefix, bucket, cache_control,
+                  aws_region, access_key_id, secret_access_key, dry_run=False):
+    '''Publishes the given directory to S3'''
+    # With glob, dotfiles are ignored by default
+    # Note that the filenames will include the `directory` prefix
+    # but we won't want that for the eventual S3 keys
+    files_and_dirs = glob.glob(path.join(directory, '**', '*'),
+                               recursive=True)
+
+    # Collect a list of all files in the specified directory
+    local_files = []
+    for filename in files_and_dirs:
+        if path.isfile(filename):
+            site_file = SiteFile(filename=filename,
+                                 dir_prefix=directory,
+                                 site_prefix=site_prefix,
+                                 cache_control=cache_control)
+            local_files.append(site_file)
+
+    # Create a list of redirects from the local files
+    local_redirects = []
+    for site_file in local_files:
+        if path.basename(site_file.filename) == 'index.html':
+            redirect_filename = path.dirname(site_file.filename)
+            site_redirect = SiteRedirect(filename=redirect_filename,
+                                         dir_prefix=directory,
+                                         site_prefix=site_prefix,
+                                         base_url=base_url)
+            local_redirects.append(site_redirect)
+
+    # Combined list of local objects
+    local_objects = local_files + local_redirects
+
+    # Create an S3 client
+    s3_client = boto3.client(
+        service_name='s3',
+        region_name=aws_region,
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key)
+
+    # Get list of remote files
+    remote_objects = list_remote_objects(bucket=bucket,
+                                         site_prefix=site_prefix,
+                                         s3_client=s3_client)
+
+    # Make dicts by filename of local and remote objects for easier searching
+    remote_objects_by_filename = {}
```

Comment:
this could be consolidated a bit:

```
remote_objects_by_filename = {
    (path.join(directory, obj.filename) if obj.filename else directory): obj
    for obj in remote_objects
}
```

Reply:
nice, thanks